src/main/java/core/basesyntax/Car.java

```diff
@@ -1,13 +1,90 @@
 package core.basesyntax;
 
+import java.util.ArrayList;
 import java.util.List;
 
-/**
- * Make this class immutable. See requirements in the README file
- */
-public class Car {
-    private int year;
-    private String color;
-    private List<Wheel> wheels;
-    private Engine engine;
+public final class Car {
+    private final int year;
+    private final String color;
+    private final List<Wheel> wheels;
+    private final Engine engine;
+
+    public Car(int year, String color, List<Wheel> wheels, Engine engine) {
+        this.year = year;
+        this.color = color;
+        this.wheels = cloneList(wheels);
+        this.engine = cloneEngine(engine);
+    }
+
+    private List<Wheel> cloneList(List<Wheel> wheels) {
+        List<Wheel> tempList = new ArrayList<>();
+        for (Wheel wheel : wheels) {
+            tempList.add(wheel == null ? null : wheel.clone());
+        }
+        return tempList;
+    }
+
+    public int getYear() {
+        return year;
+    }
+
+    public String getColor() {
+        return color;
+    }
+
+    public List<Wheel> getWheels() {
+        return cloneList(wheels);
+    }
+
+    public Engine getEngine() {
+        return cloneEngine(engine);
+    }
+
+    public Car changeEngine(Engine engine) {
+        return new Car(year, color, wheels, engine);
+    }
+
+    public Car changeColor(String newColor) {
+        return new Car(year, newColor, wheels, engine);
+    }
+
+    public Car addWheel(Wheel newWheel) {
+        List<Wheel> tempList = cloneList(wheels);
+        tempList.add(newWheel.clone());
+        return new Car(year, color, tempList, engine);
```

Comment:
```suggestion
    public Car addWheel(Wheel newWheel) {
        Car newCar = new Car(year, color, wheels, engine); 
        newCar.wheels.add(wheel.clone());
        return newCar;
```
this way you won't clone list twice